### PR TITLE
Add additional code docs controller actions: clone and destroy

### PR DESCRIPTION
**Pull request by bethanyaconnor**
This is the first PR to add an "All Code Docs" edit page. On that page, levelbuilders will be able to destroy programming environments and programming expressions, as well as clone programming expressions into other environments. These routes don't exist yet so adding them now before adding the front end to keep the PRs smaller.

## Links

[spec](https://docs.google.com/document/d/1CuMG8vlECM0xLs8kq5hocUS5ToyPWipjTGU841W2wvs/edit#bookmark=id.dwxvqefs3lom)

## Testing story

unit tests

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/programming_environments_controller.rb [hunk 1]
```diff
@@ -56,11 +56,20 @@ def update
...
   def show
-    @programming_environment = ProgrammingEnvironment.find_by_name(params[:name])
     return render :not_found unless @programming_environment
     @programming_environment_categories = @programming_environment.categories.select {|c| c.programming_expressions.count > 0}.map(&:summarize_for_environment_show)
   end
 
+  def destroy
+    return render :not_found unless @programming_environment
+    if @programming_environment.destroy
+      File.delete(@programming_environment.file_path) if File.exist?(@programming_environment.file_path)
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_environments_controller.rb]: Just checking, do we need to handle the case where `@programming_environment.destroy` succeeds but then the file delete fails?

File: dashboard/app/controllers/programming_environments_controller.rb [hunk 2]
```diff
@@ -56,11 +56,20 @@ def update
...
+      render(status: 200, plain: "Destroyed #{@programming_environment.name}")
+    else
+      render(status: :not_acceptable, plain: @programming_environment.errors.full_messages.join('. '))
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_environments_controller.rb]: (minor, up to you) This is a bit pedantic, but I think "406 Not Acceptable" is a fairly specific error code around the content-type of the response.  If this error is likely to be a server-side error, then 500 is probably more appropriate?

**Review by jamescodeorg — COMMENTED** [review]
Looks good, just a few questions.

File: dashboard/app/controllers/programming_environments_controller.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb]: We've been using "Not Acceptable" in lesson and script editing, while level uses a mis of not acceptable and unprocessable entry, which is why I've been using it here. I don't have a strong opinion personally

Comment (jamescodeorg) [reply dashboard/app/controllers/programming_environments_controller.rb]: Thanks for the context, up to you!

File: dashboard/app/controllers/programming_environments_controller.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb]: Oh that's an interesting question. I think I can solve this by using the `after_destroy` callback -- I'll move this over (for this and expressions)

Comment (jamescodeorg) [reply dashboard/app/controllers/programming_environments_controller.rb]: If we use `after_destroy`, what's the behavior if `@programming_environment.destroy` succeeds and the `File.delete` throws an error?  Is the `@programming_environment.destroy` committed to the db?  Does the client get a success or error response?  (I'm sorry for asking, my googling skills are failing me this morning.)

Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb]: I believe they're automatically wrapped in a transaction (according to the [Ruby docs](https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/Transactions/ClassMethods.html)). So my understanding is that `destroy` is not committed to the DB and this route would return a non-200 code. I'm updating this right now, so I can check that assumption

Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb]: My understanding was correct -- if the File delete fails, the object is *not* destroyed. I think this is better as it's more predictable for the DB and file system to be in sync if something did go wrong. If not, the programming environment would just get recreated on the next seed and that would be kind of tricky to track down

Comment (jamescodeorg) [reply dashboard/app/controllers/programming_environments_controller.rb]: Sounds great, thanks for digging into this!

**Pull request merged**